File: Legacy/Leetcode/78Subsets.py
# Given a set of distinct integers, nums, return all possible subsets.
#
# Note: The solution set must not contain duplicate subsets.
#
# For example,
# If nums = [1,2,3], a solution is:
#
# [
#   [3],
#   [1],
#   [2],
#   [1,2,3],
#   [1,3],
#   [2,3],
#   [1,2],
#   []
# ]

import itertools

def subsets(nums):
    """
    :type nums: List[int]
    :rtype: List[List[int]]
    """
    ans = []
    length = len(nums)
    for i in range(length + 1):
        combs = itertools.combinations(nums, i)
        for comb in combs:
            ans.append(comb)
    return ans

File: C++/Comments/Backtracking/78. Subsets.cpp
class Solution {
public:
    
    void backtrack(vector<int>& nums, vector<vector<int>>& res, vector<int>& comb, int start) {
        res.push_back(comb);
        for (int i = start; i < nums.size(); i++) {
            comb.push_back(nums[i]);
            backtrack(nums, res, comb, i+1);
            comb.pop_back();
        }
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(nums, res, comb, 0);
        return res;
    }
};